src/lib/mdns/TxtFields.cpp

```diff
@@ -0,0 +1,281 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "TxtFields.h"
+
+#include <algorithm>
+#include <cstdio>
+#include <inttypes.h>
+#include <limits>
+#include <string.h>
+
+#include <mdns/Resolver.h>
+
+namespace chip {
+namespace Mdns {
+
+namespace Internal {
+
+namespace {
+bool IsKey(const ByteSpan & key, const char * desired)
+{
+    if (key.size() != strlen(desired))
+    {
+        return false;
+    }
+    return memcmp(key.data(), desired, key.size()) == 0;
+}
+
+uint16_t MakeU16FromAsciiDecimal(const ByteSpan & val)
+{
+    uint32_t u32          = 0;
+    const uint16_t errval = 0x0;
+    for (size_t i = 0; i < val.size(); ++i)
+    {
+        char c = static_cast<char>(val.data()[i]);
+        if (c < '0' || c > '9')
+        {
+            return errval;
+        }
+        u32 = u32 * 10;
+        u32 += val.data()[i] - static_cast<uint8_t>('0');
+        if (u32 > std::numeric_limits<uint16_t>::max())
+        {
+            return errval;
+        }
+    }
+    return static_cast<uint16_t>(u32);
+}
+
+uint8_t MakeU8FromAsciiDecimal(const ByteSpan & val)
+{
+    uint16_t u16 = MakeU16FromAsciiDecimal(val);
+    if (u16 > std::numeric_limits<uint8_t>::max())
+    {
+        return 0x0;
+    }
+    return static_cast<uint8_t>(u16);
+}
+
+CHIP_ERROR MakeU8FromAsciiHex(const ByteSpan & ascii, uint8_t * val)
+{
+    if (ascii.size() != 2)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    uint8_t ret = 0;
+    for (size_t i = 0; i < ascii.size(); ++i)
+    {
+        ret    = ret << 4;
+        char c = static_cast<char>(ascii.data()[i]);
+        if (c >= '0' && c <= '9')
+        {
+            ret += ascii.data()[i] - static_cast<uint8_t>('0');
+        }
+        // Only uppercase is supported according to spec.
```

Comment:
Even though only uppercase is supported by spec, error behaviour would be much better if we supported both uppercase and lowercase. In other words, let's be strict on emitting, and lenient on receiving.

Reply:
Does the spec allow non-uppercase on receive?

Reply:
The spec is on purpose unclear about this. The usual approach to take is to be forgiving on parsing, and severe on generation.